File: 139 (Word Break)/cpp/Solution.cpp
class Solution {
public:
    bool solve(int idx, string s , unordered_set<string> &wordset,  unordered_map<int,bool> &dp){
        if (idx==s.size()) return true;
        if (dp.find(idx)!=dp.end()) return dp[idx];

        for(int i = idx+1;i<=s.size();i++){
            if (wordset.count(s.substr(idx,i-idx)) && solve(i,s,wordset,dp)){
                dp[idx] = true;
                return true;
            }
        }
        dp[idx] = false;
        return false;
        
    }

    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> wordset(wordDict.begin(), wordDict.end());
        unordered_map<int,bool> dp;
        return solve(0,s,wordset,dp);

    }
};

File: 139 (Word Break)/python/Solution.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:

        wordSet = set(wordDict)

        @cache
        def solve(idx):
            if idx==len(s): return True

            for i in range(idx+1, len(s)+1):
                if s[idx:i] in wordSet and solve(i):  return True

            return False
        
        return solve(0)